Introduced the function which clean up labels added by node-labeller

anishbista60 opened the pull request:
Removed the labels which is added to each node by virthandler

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
virt-handler does not clean up labels added by node-labeller

After this PR:
Introduced the logic to cleanup labels
<!-- (optional, in `fixes #11571(, fixes #11571, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #11571

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- https://github.com/kubevirt/kubevirt/issues/11571, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,55 @@ func DeleteAll(kv *v1.KubeVirt,
+	for _, node := range nodeList.Items {
+		labels := node.GetLabels()
+
+		kubeVietLabelPrefix := "kubevirt.io/"
```
nit: typo `kubeVietLabelPrefix ` -> `kubeVirtLabelPrefix`

alicefr replied:
You can directly use the constant in the function

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,55 @@ func DeleteAll(kv *v1.KubeVirt,
+		//If any kubevirt labels were removed , update the node labels
+		if modified {
+			node.SetLabels(labels)
+			_, err := clientset.CoreV1().Nodes().Update(context.Background(), &node, metav1.UpdateOptions{})
```
you should use `Patch` instead of `Update`

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,55 @@ func DeleteAll(kv *v1.KubeVirt,
+			}
+		}
+
+		//If any kubevirt labels were removed , update the node labels
```
nit: add a space after //

alicefr commented:
@anishbista60 thanks for the PR!
It will be good to add a couple of unit tests too.

anishbista60 commented:
@alicefr Unit testing is still to add . Apart from that , would you mind to look reamining Changes? Due to my university exam i'm bit slow , will add unit test shortly . Thank you !!

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,63 @@ func DeleteAll(kv *v1.KubeVirt,
+
+		// If any kubevirt labels were removed, update the node labels using Patch
+		if modified {
+			patchBytes, err := json.Marshal(labels)
+			if err != nil {
+				return fmt.Errorf("failed to marshal labels: %v", err)
+			}
+
+			// Construct the JSON patch operation
+			ops := fmt.Sprintf(`[{ "op": "replace", "path": "/metadata/labels", "value": %s }]`, string(patchBytes))
```
Please use the new type PatchSet introduced by https://github.com/kubevirt/kubevirt/pull/11599

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,63 @@ func DeleteAll(kv *v1.KubeVirt,
+
+	for _, node := range nodeList.Items {
+		labels := node.GetLabels()
+
```
nit: remove the newline

fossedihelm replied:
Please address this comment

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,63 @@ func DeleteAll(kv *v1.KubeVirt,
 		}
 	}
+	kubeVirtLabelPrefix := "kubevirt.io/"
+	err = deleteKubeVirtLabelsFromNodes(kv, clientset, kubeVirtLabelPrefix)
```
nit: please inline the with the error check

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,63 @@ func DeleteAll(kv *v1.KubeVirt,
 			return nil
 		}
 	}
+	kubeVirtLabelPrefix := "kubevirt.io/"
```
nit: this string isn't necessary, you can use directly the string in the function since it isn't used in any other places

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,63 @@ func DeleteAll(kv *v1.KubeVirt,
+	return nil
+}
+
+func deleteKubeVirtLabelsFromNodes(kv *v1.KubeVirt, clientset kubecli.KubevirtClient, kubeVirtLabelPrefix string) error {
```
`kv *v1.KubeVirt` param is unused. Please remove

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -463,11 +463,63 @@ func DeleteAll(kv *v1.KubeVirt,
+		}
+
+		// If any kubevirt labels were removed, update the node labels using Patch
+		if modified {
```
nit: I would prefer early exit
```
if !modified {
	continue
}
```

anishbista60 commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,13 +464,64 @@ func DeleteAll(kv *v1.KubeVirt,
 	return nil
 }
 
+func deleteKubeVirtLabelsFromNodes(clientset kubecli.KubevirtClient, kubeVirtLabelPrefix string) error {
```
@fossedihelm i passed `kubeVIrtLabelPrefix` in the function  on the recommdation of @alicefr . If you want me to drop it  , sure i'll do it .

fossedihelm replied:
I think that @alicefr wants the strings to be plain used in the for loop:
```
if strings.HasPrefix(labelkey, "kubevirt.io/") {
	delete(labels, labelkey)
	modified = true
}
```
Please keep me honest.
Anyway, if the server-side filtering is accepted this is a non-problem since for sure it will not be necessary anymore :)

anishbista60 replied:
@victortoso ok

anishbista60 commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,13 +464,64 @@ func DeleteAll(kv *v1.KubeVirt,
 
+func deleteKubeVirtLabelsFromNodes(clientset kubecli.KubevirtClient, kubeVirtLabelPrefix string) error {
+	// Get the list of nodes
+	nodeList, err := clientset.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
```
@victortoso ok . i'll wait for alicefr reply. Then apply the changes .

fossedihelm replied:
Please address this comment.

anishbista60 commented:
@alicefr if you don't mind , could you please review this code and where to add the `unit test ` ?

anishbista60 commented:
@alicefr why no one is replying on this PR . Is it already done as it was workshop demo at Kubecon ?  I'm beginner and want to contribute to kubevirt but yeah ok .

fossedihelm commented:
> @alicefr why no one is replying on this PR . Is it already done as it was workshop demo at Kubecon ? I'm beginner and want to contribute to kubevirt but yeah ok .

@anishbista60 There may be a variety of reasons why a PR remains idle for some period of time (local holidays, vacation, illness, temporarily limited capacity).
This does not mean that a PR is no longer needed; on the contrary, this one specifically fixes an open issue, so it is definitely valuable.
I know it can take a long time sometimes, but you have to be patient, the to-do list is very long. :slightly_smiling_face:

anishbista60 commented:
> Contributor

sure sir . Thanks

anishbista60 commented:
@fossedihelm sure sir thanks

fossedihelm commented:
@anishbista60 regarding the unit test, I think you can extend the already existing file https://github.com/kubevirt/kubevirt/blob/main/pkg%2Fvirt-operator%2Fresource%2Fapply%2Fdelete_test.go

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,13 +465,56 @@ func DeleteAll(kv *v1.KubeVirt,
+			return fmt.Errorf("failed to generate patch payload: %v", err)
+		}
+
+		_, err = clientset.CoreV1().Nodes().Patch(context.Background(), node.Name, types.JSONPatchType, payload, metav1.PatchOptions{})
```
nit: please inline the function with the error check

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,13 +465,55 @@ func DeleteAll(kv *v1.KubeVirt,
+			return fmt.Errorf("failed to generate patch payload: %v", err)
+		}
+
+		if err = clientset.CoreV1().Nodes().Patch(context.Background(), node.Name, types.JSONPatchType, payload, metav1.PatchOptions{}); err!=nil{
```
@anishbista60 Please run `make && make generate` and commit the changes. I think there are some missing spaces here `err!=nil{`

anishbista60 commented:
@fossedihelm could you please have a look once ?

fossedihelm commented:
/test all

anishbista60 commented:
@fossedihelm when I'm running make && make generate , it leads to modification to `bazel` file too. Till this commit I'm ignoring that changes and only commiting the changes of `delete go` . Do I need to commit both changes .

fossedihelm commented:
/test all

fossedihelm commented:
@anishbista60 Can you rebase onto the main branchand re-push everything?
`pkg/virt-operator/resource/apply/delete.go:509:12: assignment mismatch: 1 variable but clientset.CoreV1().Nodes().Patch returns 2 values`
There is a compilation error, thanks

anishbista60 commented:
@fossedihelm sorry , forget to mention issue  while  ` make generate `. Getting this error 

`ERROR: error loading package 'cmd/virt-api': Label '//vendor/kubevirt.io/client-go/version:def.bzl' is invalid because 'vendor/kubevirt.io/client-go/version' is not a package; perhaps you meant to put the colon here: '//:vendor/kubevirt.io/client-go/version/def.bzl'?
INFO: Elapsed time: 0.317s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (31 packages loaded)
    currently loading: cmd/virt-api ... (3 packages)
make: *** [Makefile:16: bazel-build] Error 1`

fossedihelm commented:
@anishbista60 All good. As already suggested, please rebase onto main, then run `make && make generate`. It should result in no error :)

anishbista60 commented:
> @anishbista60 All good. As already suggested, please rebase onto main, then run `make && make generate`. It should result in no error :)

Done , soorry i didn't see your comment .

fossedihelm commented:
> > @anishbista60 All good. As already suggested, please rebase onto main, then run `make && make generate`. It should result in no error :)
> 
> soorry i didn't see your comment . i rebased it but while running `make command ` getting permission denied
> 
> `hack/dockerized hack/build-ginkgo.sh /usr/bin/env: ‘bash\r’: Permission denied make: *** [Makefile:48: generate] Error 126`

Seems like an error in your local machine.
You could try to restart docker service and clean all the images/containers/volumes.
Just some hints (in case you use docker):
- restart docker service
- `docker system prune -a`,
- `docker volume rm (docker volume ls -q)`
Then try to run `make && make generate` again

anishbista60 commented:
@fossedihelm yeah , every thing done . could you please test it once again . i will try to go for debugging

fossedihelm commented:
/test all

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,13 +465,55 @@ func DeleteAll(kv *v1.KubeVirt,
+			return fmt.Errorf("failed to generate patch payload: %v", err)
+		}
+
+		if err = clientset.CoreV1().Nodes().Patch(context.Background(), node.Name, types.JSONPatchType, payload, metav1.PatchOptions{}); err != nil {
```
Patch returns 2 variables:

```suggestion
		if _, err = clientset.CoreV1().Nodes().Patch(context.Background(), node.Name, types.JSONPatchType, payload, metav1.PatchOptions{}); err != nil {
```

anishbista60 replied:
@fossedihelm done sir

fossedihelm commented:
/test pull-kubevirt-build

fossedihelm commented:
/test all

anishbista60 commented:
@alicefr could you please have a look once

fossedihelm commented:
> @alicefr could you please have a look once? @fossedihelm Do you have access to go for further testing?

The other tests will be executed once this PR is lgtm` ed and approved. This is how it works for every PR :)
/lgtm

alicefr commented:
This is still missing unit tests for this case, please add some in `delete_test.go`

alicefr commented:
/hold
Unit test missing

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,13 +465,55 @@ func DeleteAll(kv *v1.KubeVirt,
+			}
+		}
+
+		if !modified {
```
Instead of using modified you can simply check if the patchSet is empty, there is a function for that

anishbista60 commented:
@0xFelix @fossedihelm   could you please guide me writing unit testing cause iam new in this context.

0xFelix commented:
/test all

Looks good to me, let's see what CI says.

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +174,73 @@ var _ = Describe("Deletion", func() {
+		})
+
+		It("should delete kubevirt labels from nodes", func() {
+			nodes := &v1.NodeList{
```
```
pkg/virt-operator/resource/apply/delete_test.go:195:17: undefined: v1.NodeList
pkg/virt-operator/resource/apply/delete_test.go:196:17: undefined: v1.Node
```
You need to check your imports.
Please, always run `make`, `make generate`, and `make test` before pushing, so that you are sure it works on your local machine. Thanks

fossedihelm replied:
I don't know what's going on there in this case, but it is related to your local machine. It looks like a wrong project configuration

anishbista60 replied:
@fossedihelm now , every thing is fine .please test it .

fossedihelm replied:
@anishbista60 The only [Change](https://github.com/kubevirt/kubevirt/compare/82df0cc24d6daebebf57c1e5c5c284bd1b5c3133..9c4d92527d0b1d04575be0fcfa6cbf201df69f43) I see it related to a label string. 
The problem is that you are using `v1` as `v1 "k8s.io/api/core/v1"` but in the delete_test file is `v1 "kubevirt.io/api/core/v1"`.
I think you need to import `k8s.io/api/core/v1` maybe as (k8sv1 "k8s.io/api/core/v1") and use it `nodes := &k8sv1.NodeList{`.

But, you said that everything is fine now, are you sure you pushed all the changes? Thanks

anishbista60 replied:
@fossedihelm  Ok. Thanks

fossedihelm replied:
This is not compiling in my local machine. You have to do the same for the `v1.Node` -> `k8sv1.Node`. 
Doesn't your `make && make generate` return any errors?

anishbista60 commented:
I think i made maintainer overwhelmed due to the frequent review request on this PR . If that  the reason sorry for that.

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,73 @@ var _ = Describe("Deletion", func() {
+				},
+			}
+
+			clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[0], metav1.CreateOptions{})
+			clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[1], metav1.CreateOptions{})
+			clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[2], metav1.CreateOptions{})
```
Please check the error:
```
_, err := clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[0], metav1.CreateOptions{})
Expect(err).ToNot(HaveOccurred())
```
For all the create api call.

fossedihelm replied:
This comment is still valid

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -464,45 +465,48 @@ func DeleteAll(kv *v1.KubeVirt,
-			log.Log.Errorf(castFailedFmt, obj)
-			return nil
-		}
+	if err = deleteKubeVirtLabelsFromNodes(clientset); err != nil {
```
You are dropping `ValidatingAdmissionPolicyBinding` and `ValidatingAdmissionPolicy` cleanup. Maybe a rebase mistake

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,73 @@ var _ = Describe("Deletion", func() {
+
+		It("should delete kubevirt labels from nodes", func() {
+			nodes := &k8sv1.NodeList{
+				Items: []v1.Node{
```
`Node` is part of the `k8sv1` pkg, not `v1`

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,73 @@ var _ = Describe("Deletion", func() {
+			updatedNode1, err := clientset.CoreV1().Nodes().Get(context.Background(), "node1", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updatedNode1.Labels).ToNot(HaveKey("kubevirt.io/some-label"))
+			Expect(updatedNode1.Labels).To(HaveKey("other-labels"))
```
This comment is still valid

fossedihelm reviewed (commented):
Hey @anishbista60, there are still some comments to be addressed.
Also, I think that during the rebase you drop some valued code.
Maybe you are still getting some errors in your local environment because as is, this code will not build, since `v1.Node` does not exist and one unit test fails.
Thank you

anishbista60 commented:
@fossedihelm yes, it will do it . Please review, once i will finish everything ? There is also problem in my windows  machine.

anishbista60 commented:
@fossedihelm  got , little bit confuse., which changes should i drop ? i think first i will drop my changes , merge the main into enhance and add the changes that i dropped previously.

fossedihelm commented:
@anishbista60 You were dropping the cleanup of `validatingAdmissionPolicyBindings` and `validatingAdmissionPolicies`, which was a mistake. The cleanup that you are introducing should be in addition, and not replace any other cleanup.
Hope this helps

anishbista60 commented:
@fossedihelm in my system `make` command is not working . i'm doing it from my friend macbook . just left to run to `make file`.

fossedihelm commented:
> @fossedihelm in my system `make` command is not working . i'm doing it from my friend macbook . just left to run to `make file`.

I can highly recommend using Linux for development :) Mac is still ok (maybe with some limitations, IDK honestly).
With `make file` you mean `make generate`? Thank you.

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -498,6 +499,10 @@ func DeleteAll(kv *v1.KubeVirt,
 			log.Log.Errorf(castFailedFmt, obj)
 			return nil
 		}
+	}	
```
Empty space detected:
```suggestion
	}
```

anishbista60 replied:
@fossedihelm i didn't get it . Empty space ?

fossedihelm replied:
There is an empty space/tab char after the parenthesis. That has to be eliminated, otherwise the CI will not pass.
The `make generate` should automatically drop it for you.

anishbista60 replied:
@fossedihelm ok . Thank you

fossedihelm replied:
You are adding empty space here and then removing it in the subsequent commit.
Please remove the space addition directly here.

anishbista60 commented:
@fossedihelm yes `make generate`. my machine is window on the top of that i use linux using wsl . That why i am getting several issue. i think i have to install linux os.

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,77 @@ var _ = Describe("Deletion", func() {
+				},
+			}
+
+			_, err1 := clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[0], metav1.CreateOptions{})
+			Expect(err1).ToNot(HaveOccurred())
+			_, err2 := clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[1], metav1.CreateOptions{})
+			Expect(err2).ToNot(HaveOccurred())
+			_,err3 := clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[2], metav1.CreateOptions{})
+			Expect(err3).ToNot(HaveOccurred())
+
+
+			err := deleteKubeVirtLabelsFromNodes(kubevirtClient)
+			Expect(err).ToNot(HaveOccurred())
```
nit, reusing variable and dropping double newline:
```suggestion
			_, err := clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[0], metav1.CreateOptions{})
			Expect(err).ToNot(HaveOccurred())
			_, err = clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[1], metav1.CreateOptions{})
			Expect(err2).ToNot(HaveOccurred())
			_,err = clientset.CoreV1().Nodes().Create(context.Background(), &nodes.Items[2], metav1.CreateOptions{})
			Expect(err).ToNot(HaveOccurred())

			err = deleteKubeVirtLabelsFromNodes(kubevirtClient)
			Expect(err).ToNot(HaveOccurred())
```

fossedihelm reviewed (commented):
> @fossedihelm yes `make generate`. my machine is window on the top of that i use linux using wsl . That why i am getting several issue. i think i have to install linux os.

That's great! In the meanwhile: here two comments. I think we are really close

anishbista60 commented:
@fossedihelm done sir `make format` . Now, could you please check it once.

fossedihelm commented:
/test all

anishbista60 commented:
@fossedihelm please remove `hold`.

fossedihelm commented:
The hold is not a problem, it will not block running any tests or progressing the PR.
We will remove it once everything is ready. 
The failure is just a flake. 
/lgtm
@alicefr PTAL and feel free to unhold if everything is ok for you. Thanks 
/retest

anishbista60 commented:
@fossedihelm thanks. you're a really nice maintainer.

anishbista60 commented:
@fossedihelm would you to rerun failed cases ?

alicefr commented:
@anishbista60 can you please address the comment https://github.com/kubevirt/kubevirt/pull/11688/files#r1562074300 ?

fossedihelm commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -499,7 +499,7 @@ func DeleteAll(kv *v1.KubeVirt,
 			return nil
 		}
-	}	
+	}
```
@anishbista60 As said above, this change should be performed directly in the previous commit

fossedihelm reviewed (changes requested):
@anishbista60 Please address the comments below.
Running `make && make generate` in every single commit helps achieve clean code.

anishbista60 commented:
@fossedihelm 
The code is in proper format  in this commit `added  the unit testing` and  the commit where iam extra spacing is getting is `Introduced the function which clean up labels added by node-labeller`. `make && make format` is not resolving this issue. 
When i merged `main` into `enhance-02` during  merge conflict the bracket `}` at last  was not there maybe i deleted accidendly but it must be there inorder to complete the `for` loop. So, i added it additionally . Now ,don't know how iam getting this issue. 


```go
objects = stores.ValidatingAdmissionPolicyCache.List()
	for _, obj := range objects {
		if validatingAdmissionPolicy, ok := obj.(*admissionregistrationv1.ValidatingAdmissionPolicy); ok && validatingAdmissionPolicy.DeletionTimestamp == nil {
			if key, err := controller.KeyFunc(validatingAdmissionPolicy); err == nil {
				expectations.ValidatingAdmissionPolicy.AddExpectedDeletion(kvkey, key)
				err := clientset.AdmissionregistrationV1().ValidatingAdmissionPolicies().Delete(context.Background(), validatingAdmissionPolicy.Name, deleteOptions)
				if err != nil {
					expectations.ValidatingAdmissionPolicy.DeletionObserved(kvkey, key)
					log.Log.Errorf("Failed to delete validatingAdmissionPolicy %+v: %v", validatingAdmissionPolicy, err)
					return err
				}
			}
		} else if !ok {
			log.Log.Errorf(castFailedFmt, obj)
			return nil
		}
	}	
```

fossedihelm commented:
@anishbista60 
You need to use interctive rebase:
1. `git rebase -i HEAD~2`
2. Put an e  in the commit you want to edit so that your HEAD points to the first commit:
  ```
  e 1321805887 Introduced the function which clean up labels added by node-labeller
  pick 485d5c3765 Added unit testing for kubevirt label removal 
  ```  
3. Perform the requested changes:
   - drop the extra space at line 502 https://github.com/kubevirt/kubevirt/pull/11688#discussion_r1652389032
   - drop the extra newline at 489 https://github.com/kubevirt/kubevirt/pull/11688#discussion_r1561146919
   - using label selector at line 480 https://github.com/kubevirt/kubevirt/pull/11688#discussion_r1562074300
   Run `make and make generate`  IDK why you continue with `make format`  when I am asking for `make generate`
4. `git add -u` for adding the changes to the changelist
5. `git commit --amend --no-edit` to commit them and amend to the last commit w/o changing the message
6. `git rebase --continue` for continue the rebase (solve conflicts with the subsequent commit if requested)
7. Force push and done.

Generally, I suggest you take this as an opportunity to improve your knowledge about git with some tutorial or documentation.
Thank you!

anishbista60 commented:
@fossedihelm first thanks for your guide. i think i don't have to be  worry about rebasing stufff. I was not able to remove that extra space  . I'm so silly . i just need to take cursor one tab right from the `}` and then press backspace . It will solve the problem. 
Now, please have a look and is there any thing to add?

fossedihelm commented:
@anishbista60 Looks better, but 
- using label selector at line 480 https://github.com/kubevirt/kubevirt/pull/11688#discussion_r1562074300

still appear in the second commit, while it should be part of the first one.

anishbista60 commented:
@fossedihelm Thanks. I think now it looks good.

fossedihelm commented:
/test all

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,76 @@ var _ = Describe("Deletion", func() {
+
+		It("should delete kubevirt labels from nodes", func() {
+			nodes := &k8sv1.NodeList{
+				Items: []k8sv1.Node{
```
You need to adjust tests with `kubevirt.io/schedulable` label.
You can also add a node without it and check that it will be skipped.

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,92 @@ var _ = Describe("Deletion", func() {
+
+			updatedNode2, err := clientset.CoreV1().Nodes().Get(context.Background(), "node2", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updatedNode2.Labels).ToNot(HaveKey("kubevirt.io/another-label"))
```
node2 has 2 labels and both of them should be dropped:
```suggestion
			Expect(updatedNode2.Labels).To(BeEmpty())
```

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,92 @@ var _ = Describe("Deletion", func() {
+							Labels: map[string]string{},
+						},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "node4",
+							Labels: map[string]string{
+								"kubevirt.io/another-label": "value",
+							},
+						},
+					},
```
IMO I think you can drop this node since it is not a real case

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,92 @@ var _ = Describe("Deletion", func() {
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updatedNode3.Labels).To(BeEmpty())
+
+			updatedNode4, err := clientset.CoreV1().Nodes().Get(context.Background(), "node4", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updatedNode4.Labels).ToNot(HaveKey("kubevirt.io/another-label"))
```
And this as well

fossedihelm commented on pkg/virt-operator/resource/apply/delete_test.go:
```diff
@@ -169,6 +175,92 @@ var _ = Describe("Deletion", func() {
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updatedNode2.Labels).ToNot(HaveKey("kubevirt.io/another-label"))
+
+			updatedNode3, err := clientset.CoreV1().Nodes().Get(context.Background(), "node3", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updatedNode3.Labels).To(BeEmpty())
```
It will be good to check that no patch has been performed because it was filtered out.
Maybe adding something like this:
```
			Expect(clientset.Actions()).To(WithTransform(func(actions []testing.Action) []testing.Action {
				var node3patchActions []testing.Action
				for _, action := range actions {
					if action.GetVerb() == "patch" &&
						action.GetResource().Resource == "nodes" {
						patchAction := action.(testing.PatchAction)
						if patchAction.GetName() == "node3" {
							node3patchActions = append(node3patchActions, action)
						}
					}
				}
				return node3patchActions
			}, BeEmpty()))
```

anishbista60 replied:
@fossedihelm i made all changes based on your review.

fossedihelm reviewed (commented):
Last improvements :)

fossedihelm reviewed (approved):
Thank you

fossedihelm commented:
/test all

anishbista60 commented:
@fossedihelm What does flake mean ?

anishbista60 commented:
@fossedihelm Is this failed case like flake ?

fossedihelm commented:
@anishbista60 Yes arm unit tests is a little unstable at the moment

anishbista60 commented:
@fossedihelm could you please mention `alice` to review this PR. Thankyou

fossedihelm commented:
> @fossedihelm could you please mention `alice` to review this PR. Thankyou

@anishbista60 I think you can mention and ask for a review from Alice on your own. There is no need to request a review from another person :)  Thank you

anishbista60 commented:
@alicefr Could you please go for futher Testing ?

anishbista60 commented:
@alicefr could you please look into this PR ?

alicefr commented on pkg/virt-operator/resource/apply/delete.go:
```diff
@@ -500,13 +501,52 @@ func DeleteAll(kv *v1.KubeVirt,
 }
 
+func deleteKubeVirtLabelsFromNodes(clientset kubecli.KubevirtClient) error {
+	nodeList, err := clientset.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{LabelSelector: "kubevirt.io/schedulable"})
```
Can you please use the constant [NodeSchedulable](https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/api/core/v1/types.go#L893) instead of the hardcoded string?

anishbista60 replied:
@alicefr done

alicefr commented:
@anishbista60 just one last small comment, otherwise the rest looks good

alicefr commented:
/test all

anishbista60 commented:
@alicefr i think these failed cases is like flakes . So, could you please move for futher testing?

alicefr commented:
@anishbista60 they aren't mandatory, so don't worry about them
/approve

fossedihelm commented:
/lgtm

anishbista60 commented:
@fossedihelm could you please remove `hold` label ?

alicefr commented:
/unhold
